test/integration/stats_integration_test.cc

```diff
@@ -129,5 +136,67 @@ TEST_P(StatsIntegrationTest, WithTagSpecifierWithFixedValue) {
   EXPECT_EQ(live->tags()[0].value_, "xxx");
 }
 
+// TODO(cmluciano) Refactor once https://github.com/envoyproxy/envoy/issues/5624 is solved
+// TODO(cmluciano) Add options to measure multiple workers & without stats
+// This class itself does not add additional tests. It is a helper for use in other tests measuring
+// cluster overhead.
+class ClusterMemoryTestHelper : public BaseIntegrationTest {
+public:
+  ClusterMemoryTestHelper()
+      : BaseIntegrationTest(testing::TestWithParam<Network::Address::IpVersion>::GetParam()) {}
+
+  /**
+   *
+   * @param num_clusters number of clusters appended to bootstrap_config
+   * @param allow_stats if false, enable set_reject_all in stats_config
+   * @return size_t the total memory allocated
+   */
+  size_t ClusterMemoryHelper(int num_clusters, bool allow_stats) {
+    config_helper_.addConfigModifier([&](envoy::config::bootstrap::v2::Bootstrap& bootstrap) {
+      if (!allow_stats) {
+        bootstrap.mutable_stats_config()->mutable_stats_matcher()->set_reject_all(true);
+      }
+      for (int i = 1; i < num_clusters; i++) {
+        auto* c = bootstrap.mutable_static_resources()->add_clusters();
+        c->set_name(fmt::format("cluster_{}", i));
+      }
+    });
+    initialize();
+
+    return Memory::Stats::totalCurrentlyAllocated();
+  }
+};
+class ClusterMemoryTestRunner : public testing::TestWithParam<Network::Address::IpVersion> {};
+
+INSTANTIATE_TEST_SUITE_P(IpVersions, ClusterMemoryTestRunner,
+                         testing::ValuesIn(TestEnvironment::getIpVersionsForTest()),
+                         TestUtility::ipTestParamsToString);
+
+TEST_P(ClusterMemoryTestRunner, MemoryLargeClusterSizeWithStats) {
+  // Skip test if we cannot measure memory with TCMALLOC
+  if (!Stats::TestUtil::hasDeterministicMallocStats()) {
+    return;
+  }
+  const size_t start_mem = Memory::Stats::totalCurrentlyAllocated();
+  // A unique instance of ClusterMemoryTest allows for multiple runs of Envoy with
+  // differing configuration. This is necessary for measuring the memory consumption
+  // between the different instances within the same test.
+  auto t1 = std::make_unique<ClusterMemoryTestHelper>();
+  size_t m1 = t1->ClusterMemoryHelper(1, true);
+  EXPECT_LT(start_mem, m1);
+  t1.reset(nullptr);
```

Comment:
Nit: I would just brace scope each of the `m1/m2` tests, that way you don't need to worry about resetting things.

Reply:
I tried this out but got a stacktrace, does it look correct below 

```c++
 const size_t start_mem = Memory::Stats::totalCurrentlyAllocated();
 
  auto t1 = std::make_unique<ClusterMemoryTestHelper>();
  auto t2 = std::make_unique<ClusterMemoryTestHelper>();
            {
                size_t m1 = t1->ClusterMemoryHelper(1, true);
                EXPECT_LT(start_mem, m1);
  
                const size_t m1001 = t2->ClusterMemoryHelper(1001, true);
                EXPECT_LT(start_mem, m1001);
                const size_t m_per_cluster = (m1001 - m1) / 1000;
// As of 2019/03/13, m_per_cluster = 56404 (libstdc++), 52249 (libc++).
#ifdef _LIBCPP_VERSION
                EXPECT_LT(m_per_cluster, 53000);
#else
                EXPECT_LT(m_per_cluster, 57000);
#endif
            }
}
```

Reply:
I would have scoped instantiation of the helpers, or put the repeated block into a parameterized helper method in the helper.

```
class ClusterMemoryTestHelper  : public ... {
  ...
  static size_t computeMemory(int num_clusters) {
     const size_t start_mem = Memory::Stats::totalCurrentlyAllocated();
     ClusterMemoryTestHelper helper(num_clusters, true);
      size_t memory = helper.ClusterMemoryHelper(num_clusters, true);
      EXPECT_LT(start_mem, memory);
  }
 ```

Then in MemoryLargeClusterSizeWithStats you can just write:
```
   const size_t m1 = ClusterMemoryTestHelper::computeMemory(1);
   const size_t m1001 = ClusterMemoryTestHelper::computeMemory(1001);
   const size_t m_per_cluster = (m1001 - m1) / 1000;

   // expect statements
```
that way you solve the scoping problem and also reduce repeated code.